src/foam/dao/RoutingJDAO.js

```diff
@@ -11,75 +11,87 @@ foam.CLASS({
   requires: [
     'foam.dao.JDAO'
   ],
-  documentation:
-    `JDAO that adds the service name to the context to use for routing to correct DAO.
-    Doing this allows the underlying journal implementation to output the DAO name
-    alongside the journal entry which will aid in using a single journal file.`,
+  documentation: `JDAO that adds the service name to the context to use for
+    routing to correct DAO. Doing this allows the underlying journal
+    implementation to output the DAO name alongside the journal entry which will
+    aid in using a single journal file.`,
 
   properties: [
     {
       class: 'String',
       name: 'service',
-      documentation: 'Name of the service'
+      documentation: 'Name of the service.'
     },
     {
       class: 'FObjectProperty',
       of: 'foam.dao.DAO',
-      name: 'delegate'
+      name: 'delegate',
+      documentation: `Underlying innerDAO (e.g., MapDAO or MDAO) where the
+        entries can be directly put into on replay.`
     },
     {
       class: 'FObjectProperty',
       of: 'foam.dao.RoutingJournal',
       name: 'journal',
+      documentation: 'Shared journal where new entries are being put into.',
       javaPostSet: `
-new Thread() {
-  public void run() {
-    getJournal().waitForReplay();
-    setPromise(
-      new foam.dao.java.JDAO.Builder(getX())
-        .setOf(getOf())
-        .setDelegate(getDelegate())
-        .setJournal(getJournal())
-        .build()
-    );
-  }
-}.start();
+        new Thread() {
+          public void run() {
+            getJournal().waitForReplay();
+            setPromise(
+              new foam.dao.java.JDAO.Builder(getX())
+                .setOf(getOf())
+                .setDelegate(getDelegate())
+                .setJournal(getJournal())
+                .build()
+            );
+          }
+        }.start();
       `
     }
   ],
 
   methods: [
     {
       name: 'find_',
+      documentation: `Check if the replayingJournal is set in context. If it is,
+        then this is a find called from within a journal replay; bypass all of
+        the decorators. Otherwise, find on the decorator.`,
       flags: null,
       javaCode: `
-if ( x.get("replayingJournal") == getJournal() ) {
-  return getDelegate().find_(x, id);
-} else {
-  return super.find_(x, id);
-}
+        if ( x.get("replayingJournal") == getJournal() ) {
```

Comment:
This seems like a really odd thing to have to check for, and a really fragile condition.  Why is the default behaviour not appropriate in this case?

Reply:
While replaying, we only want DAO operations from the replaying journal to go through and any other operations should be blocked until replaying is complete.

Reply:
I read the documentation comment so I understand now.  But this still is really really fragile.

With regular journals we solve this by passing the inner dao to the journal for replay.  It would be better if we could find a way to still do that.

Perhaps in EasyDAO where we construct the RoutingJDAO, we could also do x.get("routingJournal") and give it the underlying DAO to use for the given name.  Something like x.get("routingJournal").setDAOForName(getName(), deelgate);

Reply:
@adamvy Do you still want me to fix this?

My initial plan was to use a decorator with a command and wrap around the `innerDAO`.

Reply:
Yes I'd still like to address this, I'd rather not merge the code as is as it makes me nervious :).

For the non-routing journal case, the EasyDAO or the NSpec serviceScript is responsible for making sure the journal has a reference to the underlying dao for replay purposes.  I think we can preserve this in the routing journal case.

Looks like @mcarcaso is working on something along these lines here: https://github.com/foam-framework/foam2/pull/1918/files so let's coordinate with him.